File: app/src/main/java/com/arcao/geocaching4locus/settings/fragment/filter/TerrainFilterPreferenceFragment.java
package com.arcao.geocaching4locus.settings.fragment.filter;

import android.content.SharedPreferences;
import android.os.Bundle;

import com.arcao.geocaching4locus.R;
import com.arcao.geocaching4locus.base.fragment.AbstractPreferenceFragment;
import com.arcao.geocaching4locus.settings.widget.ListPreference;

public class TerrainFilterPreferenceFragment extends AbstractPreferenceFragment {
    @Override
    public void onCreate(Bundle paramBundle) {
        super.onCreate(paramBundle);

        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preference_category_filter_terrain);
    }

    @Override
    protected void preparePreference() {
        final ListPreference terrainMinPreference = findPreference(FILTER_TERRAIN_MIN, ListPreference.class);
        final ListPreference terrainMaxPreference = findPreference(FILTER_TERRAIN_MAX, ListPreference.class);

        terrainMinPreference.setSummary(prepareRatingSummary(terrainMinPreference.getValue()));
        terrainMinPreference.setOnPreferenceChangeListener((preference, newValue) -> {
            float min = Float.parseFloat((String) newValue);
            float max = Float.parseFloat(terrainMaxPreference.getValue());

            if (min > max) {
                terrainMaxPreference.setValue((String) newValue);
                terrainMaxPreference.setSummary(prepareRatingSummary((CharSequence) newValue));
            }
            return true;
        });

        terrainMaxPreference.setSummary(prepareRatingSummary(terrainMaxPreference.getValue()));
        terrainMaxPreference.setOnPreferenceChangeListener((preference, newValue) -> {
            float min = Float.parseFloat(terrainMinPreference.getValue());
            float max = Float.parseFloat((String) newValue);

            if (min > max) {
                terrainMinPreference.setValue((String) newValue);
                terrainMinPreference.setSummary(prepareRatingSummary((CharSequence) newValue));
            }
            return true;
        });
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        super.onSharedPreferenceChanged(sharedPreferences, key);

        if (key == null)
            return;

        switch (key) {
            case FILTER_TERRAIN_MIN:
            case FILTER_TERRAIN_MAX:
                final ListPreference difficultyTerrainPreference = findPreference(key, ListPreference.class);
                difficultyTerrainPreference.setSummary(prepareRatingSummary(difficultyTerrainPreference.getEntry()));
                break;
        }
    }

    private CharSequence prepareRatingSummary(CharSequence value) {
        return preparePreferenceSummary(value, 0);
    }
}

File: app/src/main/java/com/arcao/geocaching4locus/settings/fragment/filter/TerrainFilterPreferenceFragment.kt
package com.arcao.geocaching4locus.settings.fragment.filter

import android.content.SharedPreferences
import androidx.preference.ListPreference
import com.arcao.geocaching4locus.R
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_TERRAIN_MAX
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_TERRAIN_MIN
import com.arcao.geocaching4locus.base.fragment.AbstractPreferenceFragment

class TerrainFilterPreferenceFragment : AbstractPreferenceFragment() {
    override val preferenceResource: Int
        get() = R.xml.preference_category_filter_terrain

    override fun preparePreference() {
        super.preparePreference()

        val terrainMinPreference = preference<ListPreference>(FILTER_TERRAIN_MIN)
        val terrainMaxPreference = preference<ListPreference>(FILTER_TERRAIN_MAX)

        terrainMinPreference.summary = prepareRatingSummary(terrainMinPreference.value)
        terrainMinPreference.setOnPreferenceChangeListener { _, newValue ->
            val min = (newValue as String).toFloat()
            val max = terrainMaxPreference.value.toFloat()

            if (min > max) {
                terrainMaxPreference.value = newValue
                terrainMaxPreference.summary = prepareRatingSummary(newValue as CharSequence)
            }
            true
        }

        terrainMaxPreference.summary = prepareRatingSummary(terrainMaxPreference.value)
        terrainMaxPreference.setOnPreferenceChangeListener { _, newValue ->
            val min = terrainMinPreference.value.toFloat()
            val max = (newValue as String).toFloat()

            if (min > max) {
                terrainMinPreference.value = newValue
                terrainMinPreference.summary = prepareRatingSummary(newValue as CharSequence)
            }
            true
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        super.onSharedPreferenceChanged(sharedPreferences, key)

        when (key) {
            FILTER_TERRAIN_MIN,
            FILTER_TERRAIN_MAX -> {
                preference<ListPreference>(key).apply {
                    summary = prepareRatingSummary(entry ?: return)
                }
            }
        }
    }

    private fun prepareRatingSummary(value: CharSequence): CharSequence {
        return preparePreferenceSummary(value, 0)
    }
}
